### PR TITLE
[FLINK-22085][tests] Update NetworkFailureHandler to prevent closeOnFlush() from being called recursively

### PR DESCRIPTION
**Pull request by lindong28**
## Contribution Checklist

This PR updated NetworkFailureHandler to prevent closeOnFlush() from being called recursively. The test might hang indefinitely without this change.

## Brief change log

This PR updated NetworkFailureHandler to prevent closeOnFlush() from being called recursively.

## Verifying this change

Verified that INFO level log would be printed by running `mvn test -Dtest=KafkaSourceLegacyITCase`.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

**Comment by lindong28** [comment]
Ping @dawidwys @becketqin for review.

File: flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/networking/NetworkFailureHandler.java
Comment (dawidwys) [inline flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/networking/NetworkFailureHandler.java:82]: Shouldn't we remove it only when we actually close the channel? Something like:

```
    /** Closes the specified channel after all queued write requests are flushed. */
    static void closeOnFlush(Channel channel) {
        if (channel.isConnected() && !channelsBeingClosed.contains(channel)) {
            channelsBeingClosed.add(channel);
            channel.write(ChannelBuffers.EMPTY_BUFFER).addListener(CLOSE_WITH_BOOKKEEPING);
        }
    }

    private static final ChannelFutureListener CLOSE_WITH_BOOKKEEPING =
            future ->
                    future.getChannel()
                            .close()
                            .addListener(
                                    closedChannel ->
                                            channelsBeingClosed.remove(closedChannel.getChannel()));
```

File: flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/networking/NetworkFailureHandler.java
Comment (dawidwys) [inline flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/networking/NetworkFailureHandler.java:53]: How about we use:

```
    private static final Set<Channel> channelsBeingClosed = ConcurrentHashMap.newKeySet();
```

**Review by dawidwys — CHANGES_REQUESTED** [review]
Thanks for your findings and a fix @lindong28 !

Could you @pnowojski take a look at the PR if it makes sense to you? Especially the explanation in the corresponding JIRA ticket?

File: flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/networking/NetworkFailureHandler.java
Comment (lindong28) [reply flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/networking/NetworkFailureHandler.java:82]: Thanks for the review @dawidwys!

Hmm.. I think the current approach is better because:

1) The code is less and also simpler to understand.

2) The approach makes less assumption in the sense that it just prevents closeOnFlush() from being called recursively (which we know will lead to live lock). 

In comparison, the alternative approach assumes that closeOnFlush() will never need to be called more than once for the same Channel), which may be true but is (IMO) not that obvious.

Is there any concern that we allow closeOnFlush() to be called multiple times for the same channel? If not, maybe it is better to take the simpler approach.

File: flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/networking/NetworkFailureHandler.java
Comment (lindong28) [reply flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/networking/NetworkFailureHandler.java]: Thanks @dawidwys. This is nice to know. I have updated PR as suggested.

File: flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/networking/NetworkFailureHandler.java
Comment (dawidwys) [reply flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/networking/NetworkFailureHandler.java:83]: nit: I see. Your explanation makes sense. Could we make it more explicit then that the only purpose of `channelsBeingClosed` is to prevent recursive call of `closeOnFlush`? E.g. could we move the variable right above the ` closeOnFlush`? 

Moreover how about we rephrase the comment to:

```
// The set of channels that are being closed in closeOnFlush. This is needed to prevent closing a channel recursively see FLINK-22085
```

**Comment by dawidwys** [comment]
Looks good from my side. Still I'd wait for a review from @pnowojski

**Comment by pnowojski** [comment]
I don't really remember the code, but from a quick look this change makes sense to me as well.

**Comment by dawidwys** [comment]
I will apply the change to the comment and merge the PR. Thank you @lindong28 for your contribution.

File: flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/networking/NetworkFailureHandler.java
Comment (dawidwys) [reply flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/networking/NetworkFailureHandler.java:83]: Actually @lindong28 I gave it a second thought and I think there is a problem with the current approach.

The call to `channel.write(ChannelBuffers.EMPTY_BUFFER)` is an async call. There is no guarantee it finishes when you call `channelsBeingClosed.remove(channel);`. Therefore you still may end up in a "recursive" call that you call `closeOnFlush` because of a previous `closeOnFlush` call.

We know that the call to `write` has finished only in the `ChannelFutureListener`  added via `addListener`.

Comment (lindong28) [reply flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/networking/NetworkFailureHandler.java]: Thank you @dawidwys for catching this. Yep you are right. I have updated the PR according to the suggestions.

For now I still keep the `channelsBeingClosed` above the `NetworkFailureHandler` constructor instead of moving it right above the `closeOnFlush()`.  This is to follow the current Flink code style where all `static final` variables are above the class constructor.

Could you take another look?

**Comment by lindong28** [comment]
Thank you @dawidwys for the helpful comments. I have updated the PR as suggested.

Could you take another look?

**Comment by dawidwys** [comment]
Thank you @lindong28 for your contribution!

The failure is unrelated. Merging...

**Review by dawidwys — APPROVED** [review]

**Pull request merged**
